python/ray/serve/controller.py

```diff
@@ -320,37 +323,84 @@ async def run_control_loop(self) -> None:
 
             # Update the active nodes set before updating the HTTP states, so they
             # are more consistent.
+            node_update_start_time = time.time()
             self._update_active_nodes()
+            self.node_update_duration_gauge.set(time.time() - node_update_start_time)
 
             # Don't update http_state until after the done recovering event is set,
             # otherwise we may start a new HTTP proxy but not broadcast it any
             # info about available deployments & their replicas.
             if self.http_proxy_state_manager and self.done_recovering_event.is_set():
                 try:
+                    proxy_update_start_time = time.time()
                     self.http_proxy_state_manager.update(
                         active_nodes=self._active_nodes
                     )
+                    self.proxy_update_duration_gauge.set(
+                        time.time() - proxy_update_start_time
+                    )
                 except Exception:
                     logger.exception("Exception updating HTTP state.")
 
             try:
+                dsm_update_start_time = time.time()
                 any_recovering = self.deployment_state_manager.update()
+                self.dsm_update_duration_gauge.set(time.time() - dsm_update_start_time)
                 if not self.done_recovering_event.is_set() and not any_recovering:
                     self.done_recovering_event.set()
             except Exception:
                 logger.exception("Exception updating deployment state.")
 
             try:
+                asm_update_start_time = time.time()
                 self.application_state_manager.update()
+                self.asm_update_duration_gauge.set(time.time() - asm_update_start_time)
             except Exception:
                 logger.exception("Exception updating application state.")
 
             try:
+                snapshot_start_time = time.time()
                 self._put_serve_snapshot()
+                self.snapshot_duration_gauge.set(time.time() - snapshot_start_time)
             except Exception:
                 logger.exception("Exception putting serve snapshot.")
+            loop_duration = time.time() - loop_start_time
+            if loop_duration > 60:
+                logger.warning(
+                    f"The last control loop was slow (took {loop_duration}s). "
+                    "This is likely caused by running a large number of "
+                    "replicas in a single Ray cluster. Consider using "
+                    "multiple Ray clusters."
+                )
+            self.control_loop_gauge.set(loop_duration)
             await asyncio.sleep(CONTROL_LOOP_PERIOD_S)
 
+    def _create_control_loop_metrics(self):
+        self.node_update_duration_gauge = metrics.Gauge(
```

Comment:
Should these be histograms or gauges? i'm not actually sure what best practice is here

Reply:
I chose gauges because histograms don't reveal _when_ the control loop was being particularly slow. They just show how often it was slow. Knowing when would be useful since we can try to correlate the slow-down with other events (e.g. a rolling update).

I'm not sure about the best practices either though. I'd be open to switching to histograms if there's a good reason for it.